### ISSUE TITLE
Terminology: n_jobs vs num_workers vs ncpu etc.

### ISSUE THREAD

**Issue by emmanuelle** [scikit-image/scikit-image#4876]
Since we aim at accelerating some of our functions, one way to do this is to use parallel computing. We already have one function (`restoration.cycle_spin`) which is using `dask.delayed` and several threads, there is also the attempt to parallelize `segmentation.slic` in #3120 which I plan to revive, and probably other functions will have in the future the possibility to use one or more jobs/workers (threads or processes). Therefore, I'm opening this issue so that we can decide what is the best terminology for this parameter. Should it be
- `num_workers` (as in `cycle_spin`, and as in `dask`)
- `n_jobs` (as in `scikit-learn` and in `joblib`)
- something else? (probably not, it's unfortunate enough that joblib and dask have a different convention)
I think both are fine, the choice should be more about consistency with the rest of the ecosystem: do we want to be more consistent with our backend, or with big brother scikit?

**Comment by grlee77** [comment]
I hate to say it, but there is a third contender: `workers`:

SciPy tends to use just `workers` rather than `num_workers` (e.g. in [scipy.fft functions](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fftn.html#scipy.fft.fftn), [differential_evolution](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.differential_evolution.html#scipy-optimize-differential-evolution) and [quad_vec](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.quad_vec.html#scipy-integrate-quad-vec)). It also provides a context manager that can be used to control the default number of workers, which I think is nice. See [scipy.fft.set_workers](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.set_workers.html#scipy.fft.set_workers). Adding `workers` to more functions is in their [roadmap](https://github.com/scipy/scipy/blob/eddc8385b89a09e6ed48bafaaae444c1c93d5a76/doc/source/roadmap.rst#performance-improvements).

**Comment by grlee77** [comment]
I think I actually like just `workers` the best, but don't have a problem with any of the three as long as we use it consistently within scikit-image!

**Comment by emmanuelle** [comment]
Right, `workers` also makes sense of course. And it's more a question of workers rather than jobs, so for the sake of clarity I'd prefer either `workers` or `num_workers` (not proposing `n_workers` which would be a really good name, but we don't want to multiply the number of names in the ecosystem).

So, who is for `workers` and who is for `num_workers`?

**Comment by sciunto** [comment]
> Right, `workers` also makes sense of course. And it's more a question of workers rather than jobs, 

Jobbers ? :)

**Comment by sciunto** [comment]
To be more serious, `workers` looks sufficient, but `num_workers` is more explicit, I immediately understand that I'm supposed to pass an integer.

About n_* and num_*, we have both in the lib (n_dim, n_bins, n_inliers, n_tiles) and (num_trials, num_shapes, num_peaks, num_channels, ).  A preference must be added to #2616

**Comment by alexdesiqueira** [comment]
I'd vote to maintain the consistency throughout the packages: maybe `workers` is the way to go?

**Comment by stefanv** [comment]
To be consistent with the ecosystem, this requires some careful deliberation.

@thomasjpfan wrote up a [good overview of the state of the ecosystem](https://thomasjpfan.github.io/parallelism-python-libraries-design/). I see he recommends `workers` as well.

`scikit-learn` is currently on `n_jobs`, and `scipy` on `workers`.

**Comment by lagru** [comment]
Thanks for the link. To quote from that [page](https://thomasjpfan.github.io/parallelism-python-libraries-design/)

> We use SciPy’s workers parameter because it is more consistent in controlling the number of cores used. workers denotes any form of parallelism such as: multi-threading, multiprocessing, OpenMP threads, or pthreads.

This argument works just as well in favor of `num_workers` if you replace "SciPy" with "dask". `num_workers` seems the more expressive name to me (we are not passing actual workers). But `workers` is fine with me as well and SciPy is probably closer to us than dask. I really don't care that much about which solution we settle on as long we don't stall this again for a few years. :crossed_fingers:

**Comment by stefanv** [comment]
This does seem like exactly the kind of thing we need to agree on across projects, so +1.

**Comment by stefanv** [comment]
`num_workers` (which `dask` uses) is more intuitive to me than `workers` (from SciPy); for `workers`, I'd expect to find a list of engines I can send jobs to. `dask` also uses `workers` this way (list of engines).

sklearn uses `n_jobs`, probably because of `joblib`. Followed by most of the ML learning libraries.

**Comment by thomasjpfan** [comment]
I think SciPy uses `workers` because it is not always int. For example, [scipy.integrate.quad_vec's](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.quad_vec.html) `workers` take an integer and a "map-like" callable:

![Screenshot 2024-09-05 at 12 22 04 PM](https://github.com/user-attachments/assets/8c2c3ade-61f8-4eae-b0e4-0ad29943153a)

Since scikit-learn's `n_jobs` is tied to `joblib`, I can see extending that API to actual taking a `joblib.Parallel` object.

**Comment by stefanv** [comment]
We're trying to work through this again in deciding what to do with skimage 2.0.  Perhaps we should align with at least *one* project out there; `workers` -> `int | Iterable[Worker]` (for some future definition of Worker) seems OK.
